src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -94,13 +99,27 @@ public static partial class JsonSerializer
                 }
             }
 
-            if (state.Current.ObjectState == StackFrameObjectState.MetadataRefProperty)
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataTryReadRefValue)
```

Comment:
Is it necessary to have two states here (one would suffice?). I see in other logic below they may have different paths, but it's hard to determine why.

At least we could just use an OR here instead of two branches.

Reply:
As we discussed offline with @Jozkee, we do need two states. Ultimately, finding "ref" means we can just set `state.Current.ReturnValue` to a previously seen reference (plus some validation) and exit, while finding "id" would mean we should continue reading JSON properties/read array "Values" metadata...